File: frontend_native/app/src/main/java/com/simats/formcheck/data/FormAnalyzer.kt
package com.simats.formcheck.data

import com.google.mlkit.vision.pose.Pose
import com.google.mlkit.vision.pose.PoseLandmark
import kotlin.math.abs
import kotlin.math.atan2

data class ExerciseAssessment(
    val score: Int,
    val issues: List<String>,
    val feedback: String,
    val reps: Int = 0
)

class FormAnalyzer {
    // State Tracking
    private var squatState = "UP"
    private var squatReps = 0
    private var squatLowestScore = 100

    private var pushupState = "UP"
    private var pushupReps = 0
    private var pushupLowestScore = 100

    private var lungeState = "UP"
    private var lungeReps = 0
    private var lungeLowestScore = 100

    private var jjState = "CLOSED"
    private var jjReps = 0
    private var jjLowestScore = 100

    fun analyzeSquat(pose: Pose): ExerciseAssessment {
        val landmarks = pose.allPoseLandmarks
        if (landmarks.isEmpty()) return ExerciseAssessment(0, emptyList(), "No pose detected")

        val leftHip = pose.getPoseLandmark(PoseLandmark.LEFT_HIP)
        val leftKnee = pose.getPoseLandmark(PoseLandmark.LEFT_KNEE)
        val leftAnkle = pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE)
        val leftShoulder = pose.getPoseLandmark(PoseLandmark.LEFT_SHOULDER)

        val rightHip = pose.getPoseLandmark(PoseLandmark.RIGHT_HIP)
        val rightKnee = pose.getPoseLandmark(PoseLandmark.RIGHT_KNEE)
        val rightAnkle = pose.getPoseLandmark(PoseLandmark.RIGHT_ANKLE)
        val rightShoulder = pose.getPoseLandmark(PoseLandmark.RIGHT_SHOULDER)

        if (leftHip == null || leftKnee == null || leftAnkle == null ||
            rightHip == null || rightKnee == null || rightAnkle == null ||
            leftShoulder == null || rightShoulder == null
        ) {
            return ExerciseAssessment(0, emptyList(), "Partial pose detected")
        }

        val leftKneeAngle = calculateAngle(leftHip, leftKnee, leftAnkle)
        val rightKneeAngle = calculateAngle(rightHip, rightKnee, rightAnkle)
        
        // Torso angle relative to vertical (simple proxy for "chest up")
        val torsoAngle = calculateAngle(leftShoulder, leftHip, Point(leftHip.position.x, 0f))

        val issues = mutableListOf<String>()
        val avgKneeAngle = (leftKneeAngle + rightKneeAngle) / 2

        if (avgKneeAngle > 115) {
            issues.add("Go lower")
        }

        if (torsoAngle < 45) {
            issues.add("Keep your chest up")
        }

        if (abs(leftKneeAngle - rightKneeAngle) > 20) {
            issues.add("Balance both legs")
        }

        var currentScore = 100
        currentScore -= issues.size * 20
        currentScore = currentScore.coerceIn(0, 100)

        // State Machine for Rep Counting
        if (avgKneeAngle < 115) {
            if (squatState == "UP") {
                squatState = "DOWN"
                squatLowestScore = 100 // Reset for new rep
            }
            if (currentScore < squatLowestScore) {
                squatLowestScore = currentScore
            }
        } else if (avgKneeAngle > 150) {
            if (squatState == "DOWN") {
                squatState = "UP"
                if (squatLowestScore == 100) {
                    squatReps++
                }
            }
        }

        val feedback = if (issues.isEmpty()) "Good squat form" else issues[0]

        val reportedScore = if (squatState == "DOWN") squatLowestScore else currentScore
        return ExerciseAssessment(reportedScore, issues, feedback, squatReps)
    }

    fun analyzePushup(pose: Pose): ExerciseAssessment {
        val landmarks = pose.allPoseLandmarks
        if (landmarks.isEmpty()) return ExerciseAssessment(0, emptyList(), "No pose detected")
        
        val leftShoulder = pose.getPoseLandmark(PoseLandmark.LEFT_SHOULDER)
        val leftElbow = pose.getPoseLandmark(PoseLandmark.LEFT_ELBOW)
        val leftWrist = pose.getPoseLandmark(PoseLandmark.LEFT_WRIST)
        val leftHip = pose.getPoseLandmark(PoseLandmark.LEFT_HIP)
        val leftAnkle = pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE)

        if (leftShoulder == null || leftElbow == null || leftWrist == null || leftHip == null || leftAnkle == null) {
            return ExerciseAssessment(0, emptyList(), "Partial pose detected")
        }

        val elbowAngle = calculateAngle(leftShoulder, leftElbow, leftWrist)
        val bodyAngle = calculateAngle(leftShoulder, leftHip, leftAnkle)

        val issues = mutableListOf<String>()
        if (bodyAngle < 160) issues.add("Keep your back straight")
        if (elbowAngle > 90 && elbowAngle < 150) issues.add("Go lower")

        var currentScore = 100 - (issues.size * 20)
        currentScore = currentScore.coerceIn(0, 100)
        
        // State Machine for Rep Counting
        if (elbowAngle < 90) { // Going down
            if (pushupState == "UP") {
                pushupState = "DOWN"
                pushupLowestScore = 100
            }
            if (currentScore < pushupLowestScore) {
                pushupLowestScore = currentScore
            }
        } else if (elbowAngle > 150) { // Coming up
            if (pushupState == "DOWN") {
                pushupState = "UP"
                if (pushupLowestScore == 100) {
                    pushupReps++
                }
            }
        }

        val reportedScore = if (pushupState == "DOWN") pushupLowestScore else currentScore
        return ExerciseAssessment(reportedScore, issues, if (issues.isEmpty()) "Good push-up form" else issues[0], pushupReps)
    }

    fun analyzePlank(pose: Pose): ExerciseAssessment {
        val landmarks = pose.allPoseLandmarks
        if (landmarks.isEmpty()) return ExerciseAssessment(0, emptyList(), "No pose detected")

        val shoulder = pose.getPoseLandmark(PoseLandmark.LEFT_SHOULDER)
        val hip = pose.getPoseLandmark(PoseLandmark.LEFT_HIP)
        val ankle = pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE)

        if (shoulder == null || hip == null || ankle == null) {
            return ExerciseAssessment(0, emptyList(), "Partial pose detected")
        }

        val bodyAngle = calculateAngle(shoulder, hip, ankle)
        val issues = mutableListOf<String>()
        
        if (bodyAngle < 165) issues.add("Lower your hips")
        else if (bodyAngle > 195) issues.add("Raise your hips")

        var score = 100 - (issues.size * 20)
        score = score.coerceIn(0, 100)
        return ExerciseAssessment(score, issues, if (issues.isEmpty()) "Good plank form" else issues[0])
    }

    fun analyzeLunge(pose: Pose): ExerciseAssessment {
        val landmarks = pose.allPoseLandmarks
        if (landmarks.isEmpty()) return ExerciseAssessment(0, emptyList(), "No pose detected")

        val leftHip = pose.getPoseLandmark(PoseLandmark.LEFT_HIP)
        val leftKnee = pose.getPoseLandmark(PoseLandmark.LEFT_KNEE)
        val leftAnkle = pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE)
        
        if (leftHip == null || leftKnee == null || leftAnkle == null) {
            return ExerciseAssessment(0, emptyList(), "Partial pose detected")
        }

        val kneeAngle = calculateAngle(leftHip, leftKnee, leftAnkle)
        val issues = mutableListOf<String>()

        if (kneeAngle > 110) issues.add("Drop your back knee lower")

        var currentScore = 100 - (issues.size * 20)
        currentScore = currentScore.coerceIn(0, 100)

        // State Machine for Rep Counting
        if (kneeAngle < 100) {
            if (lungeState == "UP") {
                lungeState = "DOWN"
                lungeLowestScore = 100
            }
            if (currentScore < lungeLowestScore) {
                lungeLowestScore = currentScore
            }
        } else if (kneeAngle > 150) {
            if (lungeState == "DOWN") {
                lungeState = "UP"
                if (lungeLowestScore == 100) {
                    lungeReps++
                }
            }
        }

        val reportedScore = if (lungeState == "DOWN") lungeLowestScore else currentScore
        return ExerciseAssessment(reportedScore, issues, if (issues.isEmpty()) "Good lunge form" else issues[0], lungeReps)
    }

    fun analyzeJumpingJack(pose: Pose): ExerciseAssessment {
        val landmarks = pose.allPoseLandmarks
        if (landmarks.isEmpty()) return ExerciseAssessment(0, emptyList(), "No pose detected")

        val leftAnkle = pose.getPoseLandmark(PoseLandmark.LEFT_ANKLE)
        val rightAnkle = pose.getPoseLandmark(PoseLandmark.RIGHT_ANKLE)
        val leftWrist = pose.getPoseLandmark(PoseLandmark.LEFT_WRIST)
        val rightWrist = pose.getPoseLandmark(PoseLandmark.RIGHT_WRIST)

        if (leftAnkle == null || rightAnkle == null || leftWrist == null || rightWrist == null) {
            return ExerciseAssessment(0, emptyList(), "Partial pose detected")
        }

        // Just basic checks for a jumping jack
        val feetDistance = abs(leftAnkle.position.x - rightAnkle.position.x)
        val handsDistance = abs(leftWrist.position.x - rightWrist.position.x)

        val issues = mutableListOf<String>()
        if (feetDistance < 50 && handsDistance > 100) issues.add("Coordinate arms and legs")

        var currentScore = 100 - (issues.size * 20)
        currentScore = currentScore.coerceIn(0, 100)

        // State Machine for Rep Counting
        if (handsDistance > 150 && feetDistance > 100) {
            if (jjState == "CLOSED") {
                jjState = "OPEN"
                jjLowestScore = 100
            }
            if (currentScore < jjLowestScore) {
                jjLowestScore = currentScore
            }
        } else if (handsDistance < 100 && feetDistance < 50) {
            if (jjState == "OPEN") {
                jjState = "CLOSED"
                if (jjLowestScore == 100) {
                    jjReps++
                }
            }
        }

        val reportedScore = if (jjState == "OPEN") jjLowestScore else currentScore
        return ExerciseAssessment(reportedScore, issues, if (issues.isEmpty()) "Good jumping jack form" else issues[0], jjReps)
    }

    private fun calculateAngle(firstPoint: PoseLandmark, midPoint: PoseLandmark, lastPoint: PoseLandmark): Float {
        var result = java.lang.Math.toDegrees(
            (atan2(lastPoint.position.y - midPoint.position.y, lastPoint.position.x - midPoint.position.x) -
                    atan2(firstPoint.position.y - midPoint.position.y, firstPoint.position.x - midPoint.position.x)).toDouble()
        ).toFloat()
        result = abs(result)
        if (result > 180) {
            result = 360.0f - result
        }
        return result
    }

    private fun calculateAngle(firstPoint: PoseLandmark, midPoint: PoseLandmark, lastPoint: Point): Float {
        var result = java.lang.Math.toDegrees(
            (atan2(lastPoint.y - midPoint.position.y, lastPoint.x - midPoint.position.x) -
                    atan2(firstPoint.position.y - midPoint.position.y, firstPoint.position.x - midPoint.position.x)).toDouble()
        ).toFloat()
        result = abs(result)
        if (result > 180) {
            result = 360.0f - result
        }
        return result
    }

    private data class Point(val x: Float, val y: Float)
}

File: web/src/lib/FormAnalyzer.js
class FormAnalyzer {
  constructor() {
    this.squatState = "UP";
    this.squatReps = 0;
    this.squatLowestScore = 100;

    this.pushupState = "UP";
    this.pushupReps = 0;
    this.pushupLowestScore = 100;

    this.lungeState = "UP";
    this.lungeReps = 0;
    this.lungeLowestScore = 100;

    this.jjState = "CLOSED";
    this.jjReps = 0;
    this.jjLowestScore = 100;
  }

  getLandmark(keypoints, name) {
    const point = keypoints.find(k => k.name === name);
    return point && point.score > 0.3 ? point : null;
  }

  calculateAngle(firstPoint, midPoint, lastPoint) {
    let result = Math.atan2(lastPoint.y - midPoint.y, lastPoint.x - midPoint.x) -
                 Math.atan2(firstPoint.y - midPoint.y, firstPoint.x - midPoint.x);
    result = Math.abs((result * 180.0) / Math.PI);
    if (result > 180.0) {
      result = 360.0 - result;
    }
    return result;
  }

  analyzeSquat(keypoints) {
    if (!keypoints || keypoints.length === 0) return { score: 0, issues: ["No pose detected"], feedback: "No pose detected", reps: this.squatReps };

    const leftHip = this.getLandmark(keypoints, 'left_hip');
    const leftKnee = this.getLandmark(keypoints, 'left_knee');
    const leftAnkle = this.getLandmark(keypoints, 'left_ankle');
    const leftShoulder = this.getLandmark(keypoints, 'left_shoulder');

    const rightHip = this.getLandmark(keypoints, 'right_hip');
    const rightKnee = this.getLandmark(keypoints, 'right_knee');
    const rightAnkle = this.getLandmark(keypoints, 'right_ankle');
    const rightShoulder = this.getLandmark(keypoints, 'right_shoulder');

    if (!leftHip || !leftKnee || !leftAnkle || !rightHip || !rightKnee || !rightAnkle || !leftShoulder || !rightShoulder) {
      return { score: 0, issues: ["Partial pose detected"], feedback: "Partial pose detected", reps: this.squatReps };
    }

    const leftKneeAngle = this.calculateAngle(leftHip, leftKnee, leftAnkle);
    const rightKneeAngle = this.calculateAngle(rightHip, rightKnee, rightAnkle);
    const torsoAngle = this.calculateAngle(leftShoulder, leftHip, { x: leftHip.x, y: 0 });

    const issues = [];
    const avgKneeAngle = (leftKneeAngle + rightKneeAngle) / 2;

    if (avgKneeAngle > 115) issues.push("Go lower");
    if (torsoAngle < 45) issues.push("Keep your chest up");
    if (Math.abs(leftKneeAngle - rightKneeAngle) > 20) issues.push("Balance both legs");

    let currentScore = 100 - (issues.length * 20);
    currentScore = Math.max(0, Math.min(100, currentScore));

    if (avgKneeAngle < 115) {
      if (this.squatState === "UP") {
        this.squatState = "DOWN";
        this.squatLowestScore = 100;
      }
      if (currentScore < this.squatLowestScore) {
        this.squatLowestScore = currentScore;
      }
    } else if (avgKneeAngle > 150) {
      if (this.squatState === "DOWN") {
        this.squatState = "UP";
        if (this.squatLowestScore === 100) {
          this.squatReps++;
        }
      }
    }

    const feedback = issues.length === 0 ? "Good squat form" : issues[0];
    const reportedScore = this.squatState === "DOWN" ? this.squatLowestScore : currentScore;
    
    return { score: reportedScore, issues, feedback, reps: this.squatReps };
  }

  analyzePushup(keypoints) {
    if (!keypoints || keypoints.length === 0) return { score: 0, issues: ["No pose detected"], feedback: "No pose detected", reps: this.pushupReps };

    const leftShoulder = this.getLandmark(keypoints, 'left_shoulder');
    const leftElbow = this.getLandmark(keypoints, 'left_elbow');
    const leftWrist = this.getLandmark(keypoints, 'left_wrist');
    const leftHip = this.getLandmark(keypoints, 'left_hip');
    const leftAnkle = this.getLandmark(keypoints, 'left_ankle');

    if (!leftShoulder || !leftElbow || !leftWrist || !leftHip || !leftAnkle) {
      return { score: 0, issues: ["Partial pose detected"], feedback: "Partial pose detected", reps: this.pushupReps };
    }

    const elbowAngle = this.calculateAngle(leftShoulder, leftElbow, leftWrist);
    const bodyAngle = this.calculateAngle(leftShoulder, leftHip, leftAnkle);

    const issues = [];
    if (bodyAngle < 160) issues.push("Keep your back straight");
    if (elbowAngle > 90 && elbowAngle < 150) issues.push("Go lower");

    let currentScore = 100 - (issues.length * 20);
    currentScore = Math.max(0, Math.min(100, currentScore));

    if (elbowAngle < 90) {
      if (this.pushupState === "UP") {
        this.pushupState = "DOWN";
        this.pushupLowestScore = 100;
      }
      if (currentScore < this.pushupLowestScore) {
        this.pushupLowestScore = currentScore;
      }
    } else if (elbowAngle > 150) {
      if (this.pushupState === "DOWN") {
        this.pushupState = "UP";
        if (this.pushupLowestScore === 100) {
          this.pushupReps++;
        }
      }
    }

    const feedback = issues.length === 0 ? "Good push-up form" : issues[0];
    const reportedScore = this.pushupState === "DOWN" ? this.pushupLowestScore : currentScore;
    
    return { score: reportedScore, issues, feedback, reps: this.pushupReps };
  }

  analyzePlank(keypoints) {
    if (!keypoints || keypoints.length === 0) return { score: 0, issues: ["No pose detected"], feedback: "No pose detected", reps: 0 };

    const shoulder = this.getLandmark(keypoints, 'left_shoulder');
    const hip = this.getLandmark(keypoints, 'left_hip');
    const ankle = this.getLandmark(keypoints, 'left_ankle');

    if (!shoulder || !hip || !ankle) {
      return { score: 0, issues: ["Partial pose detected"], feedback: "Partial pose detected", reps: 0 };
    }

    const bodyAngle = this.calculateAngle(shoulder, hip, ankle);
    const issues = [];

    if (bodyAngle < 165) issues.push("Lower your hips");
    else if (bodyAngle > 195) issues.push("Raise your hips");

    let currentScore = 100 - (issues.length * 20);
    currentScore = Math.max(0, Math.min(100, currentScore));

    const feedback = issues.length === 0 ? "Good plank form" : issues[0];
    return { score: currentScore, issues, feedback, reps: 0 };
  }

  analyzeLunge(keypoints) {
    if (!keypoints || keypoints.length === 0) return { score: 0, issues: ["No pose detected"], feedback: "No pose detected", reps: this.lungeReps };

    const leftHip = this.getLandmark(keypoints, 'left_hip');
    const leftKnee = this.getLandmark(keypoints, 'left_knee');
    const leftAnkle = this.getLandmark(keypoints, 'left_ankle');

    if (!leftHip || !leftKnee || !leftAnkle) {
      return { score: 0, issues: ["Partial pose detected"], feedback: "Partial pose detected", reps: this.lungeReps };
    }

    const kneeAngle = this.calculateAngle(leftHip, leftKnee, leftAnkle);
    const issues = [];

    if (kneeAngle > 110) issues.push("Drop your back knee lower");

    let currentScore = 100 - (issues.length * 20);
    currentScore = Math.max(0, Math.min(100, currentScore));

    if (kneeAngle < 100) {
      if (this.lungeState === "UP") {
        this.lungeState = "DOWN";
        this.lungeLowestScore = 100;
      }
      if (currentScore < this.lungeLowestScore) {
        this.lungeLowestScore = currentScore;
      }
    } else if (kneeAngle > 150) {
      if (this.lungeState === "DOWN") {
        this.lungeState = "UP";
        if (this.lungeLowestScore === 100) {
          this.lungeReps++;
        }
      }
    }

    const feedback = issues.length === 0 ? "Good lunge form" : issues[0];
    const reportedScore = this.lungeState === "DOWN" ? this.lungeLowestScore : currentScore;

    return { score: reportedScore, issues, feedback, reps: this.lungeReps };
  }

  analyzeJumpingJack(keypoints) {
    if (!keypoints || keypoints.length === 0) return { score: 0, issues: ["No pose detected"], feedback: "No pose detected", reps: this.jjReps };

    const leftAnkle = this.getLandmark(keypoints, 'left_ankle');
    const rightAnkle = this.getLandmark(keypoints, 'right_ankle');
    const leftWrist = this.getLandmark(keypoints, 'left_wrist');
    const rightWrist = this.getLandmark(keypoints, 'right_wrist');

    if (!leftAnkle || !rightAnkle || !leftWrist || !rightWrist) {
      return { score: 0, issues: ["Partial pose detected"], feedback: "Partial pose detected", reps: this.jjReps };
    }

    const feetDistance = Math.abs(leftAnkle.x - rightAnkle.x);
    const handsDistance = Math.abs(leftWrist.x - rightWrist.x);
    
    const issues = [];
    if (feetDistance < 50 && handsDistance > 100) issues.push("Coordinate arms and legs");

    let currentScore = 100 - (issues.length * 20);
    currentScore = Math.max(0, Math.min(100, currentScore));

    if (handsDistance > 150 && feetDistance > 100) {
      if (this.jjState === "CLOSED") {
        this.jjState = "OPEN";
        this.jjLowestScore = 100;
      }
      if (currentScore < this.jjLowestScore) {
        this.jjLowestScore = currentScore;
      }
    } else if (handsDistance < 100 && feetDistance < 50) {
      if (this.jjState === "OPEN") {
        this.jjState = "CLOSED";
        if (this.jjLowestScore === 100) {
          this.jjReps++;
        }
      }
    }

    const feedback = issues.length === 0 ? "Good jumping jack form" : issues[0];
    const reportedScore = this.jjState === "OPEN" ? this.jjLowestScore : currentScore;

    return { score: reportedScore, issues, feedback, reps: this.jjReps };
  }
}

export default new FormAnalyzer();
